src/flamenco/runtime/fd_blockstore.c

```diff
@@ -1045,13 +1076,14 @@ fd_blockstore_block_data_query_volatile( fd_blockstore_t * blockstore, ulong slo
      destination data to be overwritten/invalid as long as the memory
      location is valid. As long as we don't crash, we can validate the
      data after it is read. */
-  fd_wksp_t * wksp = fd_blockstore_wksp( blockstore );
-  fd_block_map_t const * block_map = fd_wksp_laddr_fast( wksp, blockstore->slot_map_gaddr );
+  fd_wksp_t *                wksp           = fd_blockstore_wksp( blockstore );
+  fd_block_map_map_t const * block_map      = fd_blockstore_block_map( blockstore );
+  fd_block_map_t     const * block_map_pool = fd_blockstore_block_map_pool( blockstore ); 
   for(;;) {
     uint seqnum;
     if( FD_UNLIKELY( fd_readwrite_start_concur_read( &blockstore->lock, &seqnum ) ) ) continue;
 
-    fd_block_map_t const * query = fd_block_map_query_safe( block_map, &slot, NULL );
+    fd_block_map_t const * query = fd_block_map_map_ele_query_const( block_map, &slot, NULL, block_map_pool );
```

Comment:
`query_const` is actually not equivalent to `query_safe`. it's not safe if there are concurrent write ops (insert, remove, query)

we'll need to add a `query_volatile` to `fd_map_chain.c`

Reply:
https://github.com/firedancer-io/firedancer/pull/2800